### PR TITLE
feat(tasks): add a tasks functionality

### PR DESCRIPTION
**Pull request by jnizet**
WIP!

Yesterday at the demo with Malika, from Globe, one of the things that appeared is that they need some kind of todo list, potentially with reminders, because 

 - they currently pass each other hand-written undecipherable notes, and 
 - they need to remember to check some stuff (like reminding a person to go to an administrative or medical appointment, etc.)

So I suggested to make something very generic, that could be very flexible but fill their needs: a simple task list, where each task could have a due date (allowing reminders, alerts, etc.), optionally a concerned person, and completely free text describing the task to do.

I though being able to assign a task to someone (or to self) would also be useful.

This is a first PR showing how the todo list could look like. It's real, but it doesn't have any screen  to edit the tasks yet, and the actions (assign, , delete, etc.) don't do anything yet.

Please tell me what you think. In particular, @acrepet, did I understand the need correctly, and do you think something important is missing?

Here's a script you can execute to populate the task table with a few examples:

```
insert into task (id, description, title, status, due_date, creator_id, assignee_id, concerned_person_id)
    VALUES (1, E'Il faudrait faire un truc\navec une description\n sur plusieurs lignes', 'Truc à faire', 'TODO', '2017-07-28', 1, 1, null);
insert into task (id, description, title, status, due_date, creator_id, assignee_id, concerned_person_id)
    VALUES (2, 'Il faudrait faire un autre truc', 'Autre truc à faire', 'TODO', '2017-08-03', 1, null, null);
insert into task (id, description, title, status, due_date, creator_id, assignee_id, concerned_person_id)
    VALUES (3, 'Et même peut-être autre chose', 'Et même peut-être autre chose', 'TODO', null, 1, null, null);
```

And here are two screenshots showing the feature:

capture d ecran 2017-07-29 a 12 28 49

capture d ecran 2017-07-29 a 12 20 04

### CODE REVIEW

File: frontend/src/app/app.module.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import { BrowserModule } from '@angular/platform-browser';
-import { LOCALE_ID, NgModule } from '@angular/core';
+import { Inject, LOCALE_ID, NgModule } from '@angular/core';
```
Comment (cexbrayat) [inline frontend/src/app/app.module.ts]: Unneeded I guess

File: frontend/src/app/app.routes.ts [hunk 2]
```diff
@@ -156,6 +169,53 @@ export const routes: Routes = [
...
+            path: 'unassigned',
+            component: TasksComponent,
+            data: {
+              taskListType: 'unassigned'
+            },
+            resolve: {
+              tasks: TasksResolverService
+            }
+          }
+        ]
```
Comment (cexbrayat) [inline frontend/src/app/app.routes.ts]: Any pros to have 4 specific routes instead of a path or a query param?

Comment (jnizet) [inline frontend/src/app/app.routes.ts]: You're right, it's probably a better idea

**Comment by jnizet** [comment]
Regarding done tasks, yes, it's in my plans. I also wonder if we really should delete tasks (even with a confirmation), or if we should simply mark them cancelled. We could then have an "Old" tab, with pagination, showing the tasks that are done or cancelled, and could propose to resurrect them. We would probably need to add a closingDate field, to be able to list the recently closed tasks first.

This can still change but I wanted to show a quick prototype to see if Agnès was on the same page.

File: frontend/src/app/app.routes.ts (hunk 2 above)
Comment (jnizet) [inline frontend/src/app/app.routes.ts]: Actually, there's something I would lose with params: there doesn't seem to be a way to have a root `/tasks` route that would redirect to `/tasks;tab=todo`, for example. So I think I'll leave it as is for the moment.

**Comment by acrepet** [comment]
OK. I test this PR. Perfect job as usual @jnizet !

- _Done tasks_ could be a nice feature
- a Delete or "Cancel" function to "not physically" delete tasks could be nice. I'm like @cexbrayat I very often fuck up and close the wrong task ;-)
- an "Old" tab, with pagination, could be nice (I just hope that there will not have so many tasks) and propose to resurrect them could be perfect. 
- a Pending function to put the task on pending

File: frontend/src/app/tasks/tasks.component.html [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      <small *ngIf="task.model.dueDate" [class]="task.dueDateClass()">{{ task.relativeDueDate() }}</small>
+    </div>
+    <div *ngIf="task.opened" class="task-description">{{ task.model.description }}</div>
+    <div class="d-flex w-100 justify-content-between align-items-start">
+      <div>
+        <small class="text-muted" *ngIf="task.model.creator">Créée par {{ task.model.creator.login }}.</small>
+        <small class="text-muted" *ngIf="task.model.assignee">Assignée à {{ task.model.assignee.login }}.</small>
+        <small class="text-muted" *ngIf="task.model.concernedPerson">Concerne <a [routerLink]="['/persons', task.model.concernedPerson.id]">{{ task.model.concernedPerson | fullname }}</a>.</small>
+      </div>
+      <div class="task-buttons">
```
Comment (acrepet) [inline frontend/src/app/tasks/tasks.component.html]: it could nice perhaps to have Tooltips on buttons

File: frontend/src/app/app.routes.ts [hunk 4]
```diff
@@ -156,9 +169,35 @@ export const routes: Routes = [
...
         path: 'cities',
         component: CitiesUploadComponent
       }
     ]
   }
 ];
+
+/**
+ * Creates a task route of a given type
+ * This function has no readon to be exported, other than to make the AOT compiler happy.
```
Comment (acrepet) [inline frontend/src/app/app.routes.ts]: readon -> reason

File: backend/src/main/java/org/globe42/domain/Task.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    private Long id;
+
+    /**
+     * The description of the task, which is free text describing what to do.
+     */
+    @NotEmpty
+    private String description;
+
+    /**
+     * A one-line title shortly describing th task, which appears in the to-do list.
```
Comment (acrepet) [inline backend/src/main/java/org/globe42/domain/Task.java]: th -> the

File: frontend/src/app/tasks/tasks.component.html (hunk 3 above)
Comment (jnizet) [inline frontend/src/app/tasks/tasks.component.html]: Yes, I tried adding some, but there is apparently a positioning bug with tooltips, which makes them look crappy. I could add a simple title attribute, but I thought I'd just wait a bit.

**Comment by jnizet** [comment]
I have added a paginated archived tasks list.

Regarding pending tasks, I'd prefer to leave that for later, if needed. Let's keep things simple for a first version.

**Comment by acrepet** [comment]
Regarding pending tasks, no problem. It's perhaps overkill

**Comment by jnizet** [comment]
Now that I think about it, to make things simpler, I think I should paginate all the tasks pages. What do you think?

**Comment by acrepet** [comment]
Yes! It could be great

**Pull request merged**